src/Modules/DataIO/ReadFile.cc

```diff
@@ -45,10 +51,193 @@ void ReadFile::setStateDefaults()
+  template <typename DataHandle>
+  using ReadFuncList = std::vector<ReadFunc<DataHandle>>;
+  template <typename DataHandle>
+  using PrioritizedReaderMap = std::map<FileExtension, ReadFuncList<DataHandle>>;
+
+  //TODO
+  //class HasLogger
+  static LoggerHandle nullLogger;
+
+  template <class DataHandle>
+  class PluginReader
+  {
+  public:
+    explicit PluginReader(const std::string& pluginFileTypeName) : pluginFileTypeName_(pluginFileTypeName) {}
+    DataHandle operator()(const Filename& filename) const
+    {
+      GenericIEPluginManager<typename DataHandle::element_type> mgr;
+      auto pl = mgr.get_plugin(pluginFileTypeName_);
+      if (pl)
+        return pl->readFile(filename.string(), nullLogger);
+      THROW_ALGORITHM_INPUT_ERROR_SIMPLE("No custom reader found to handle that file type.");
+    }
+  private:
+    std::string pluginFileTypeName_;
+  };
+
+  template <class DataHandle>
+  class BuiltInReader
+  {
+  public:
+    DataHandle operator()(const Filename& name) const
+    {
+      auto filename = name.string();
+      auto stream = auto_istream(filename, nullLogger);
+      if (!stream)
+      {
+        THROW_ALGORITHM_INPUT_ERROR_SIMPLE("Error reading file '" + filename + "'.");
+      }
+
+      DataHandle handle;
+      Pio(*stream, handle);
+
+      if (!handle || stream->error())
+      {
+        THROW_ALGORITHM_INPUT_ERROR_SIMPLE("Error reading data from file '" + filename + "'.");
+      }
+      return handle;
+    }
+  };
+
+  template <class DataHandle>
+  struct PrioritizedReaderMapBuilder
+  {
+    static PrioritizedReaderMap<DataHandle> value;
+  };
+
+  template <class DataHandle>
+  class QuickReader
+  {
+  public:
+    DataHandle read(const Filename& filename) const
+    {
+      auto ext = filename.extension().string();
+      auto readerListIter = prioritizedReaderMap.find(ext);
+      if (readerListIter != prioritizedReaderMap.end())
+      {
+        for (const auto& reader : readerListIter->second)
+        {
+          auto val = reader(filename);
+          if (val)
+            return val;
+        }
+        THROW_ALGORITHM_INPUT_ERROR_SIMPLE("No available reader could handle the file type.");
+      }
+      THROW_ALGORITHM_INPUT_ERROR_SIMPLE("No reader found to handle that file type.");
+    }
+  private:
+    PrioritizedReaderMap<DataHandle> prioritizedReaderMap = PrioritizedReaderMapBuilder<DataHandle>::value;
+  };
+
+  template <>
+  PrioritizedReaderMap<MatrixHandle> PrioritizedReaderMapBuilder<MatrixHandle>::value =
+  {
+    // .mat type: try reading as Matlab first, then SCIRun format
+    {".mat", { PluginReader<MatrixHandle>("Matlab Matrix"), BuiltInReader<MatrixHandle>() }},
+    {".igb", { PluginReader<MatrixHandle>("IGBFile") }},
+    {".txt", { PluginReader<MatrixHandle>("SimpleTextFile") }},
+    {"", { PluginReader<MatrixHandle>("ECGSimFile"),
+          PluginReader<MatrixHandle>("ECGSimFileBinary"),
+          PluginReader<MatrixHandle>("SimpleTextFile") }},
```

Comment:
@jessdtate What should the read order be for non-extension matrix files, SimpleTextFile first, then ECG?